Improve hash-to-string conversion performance

afilip1 opened the pull request:
Hi there,

Noticed some inefficiencies in string conversion code and tried to improve it a bit. In my highly scientific™ test results it performs about 3-3.5x faster on average.

Test results of converting 84 hashes of real data 2000 times:

```
Number of hashes: 84
Hash matrix dimensions: 8x8
Original version: 16.906992000000173
Updated version: 5.23041590000048
Encoded hex strings are identical:  True
```

I opted for an f-string which might raise some backwards compatibility concerns, let me know if you want it changed. Other than that I don't think this should cause any regressions.

Hope it helps.

afilip1 commented:
Apologies for the commit kerfuffle, in my infinite wisdom I forgot to account for the leading zeroes. Also adjusted formatting for 2.x compat.

Running the same benchmark (5000 iterations):
```
Original version: 41.99307150000095
Updated version: 13.63205729999936
```

JohannesBuchner commented:
Would this work OK with very large hashes (e.g., 64x64)?

JohannesBuchner commented on imagehash.py:
```diff
@@ -71,14 +71,18 @@
-	return '{:0>{width}x}'.format(int(bit_string, 2), width=width)
+	hash_int = 0
+	for bit in map(int, numpy.nditer(arr)):
+		hash_int = hash_int << 1 | bit
```
How about
`hash_int = sum(2**bit for bit in numpy.nditer(arr.flatten()).astype(int))`
or even
`hash_int = (2**numpy.nditer(arr.flatten()).astype(int))).sum()`

afilip1 replied:
I'm afraid I don't see how those options could yield a correct result. If I'm not mistaken, `2**bit` would either yield either a 1 or a 2, which I think is not what you intend.

I suspect you mean to have the digit places in `bit`, which is indeed a good option to explore, however I can't think of a good succinct way to inject those values in there off the top of my head at the moment. I'm also slightly cautious about the use of exponentiation with regards to its effect on performance, as I'm not an expert on Python specifics.

Bit shifting is about as simple and fast as it gets, and although it's not a nice one-liner, I'd say it's very readable in this case.

JohannesBuchner replied:
Hmm, you are right, my suggestion is incorrect. I was just thinking that any numpy vectorization should be faster than python iterations. Perhaps something like:
`hash_int = (2**numpy.arange(arr.size).reshape(arr.shape) * arr)).sum()`

afilip1 replied:
That was my intuition at first too, however I found out that since numpy uses fixed width integers internally, for any matrix above around 8 it starts overflowing and producing garbage results. The proposed method relies on python's infinite precision ints, and I can't really think of a way to shift it over into C.

JohannesBuchner commented on imagehash.py:
```diff
@@ -89,7 +93,7 @@ def __init__(self, binary_array):
 
 	def __str__(self):
-		return _binary_array_to_hex(self.hash.flatten())
+		return _binary_array_to_hex(self.hash)
```
why flatten here instead of in the other function?

afilip1 replied:
> Would this work OK with very large hashes (e.g., 64x64)?

Yes, it seems to keep the advantage on large hashes, too. Converting 3000 randomly generated 64x64 hashes gives the following results:
```
Original version: 15.335927000000083
Updated version: 7.23339599999963
```

It only gives in at about 160-dimensional hashes (25000+ bits). Running 1000 160x160 hashes gives a rough parity:
```
Original version: 37.50718020000204
Updated version: 37.62010579999696
```

afilip1 replied:
I found that `nditer()` function iterates over all elements of multidimensional arrays already, so calling `flatten()` is not necessary here, or perhaps even harmful due to the cost of array copying.

afilip1 commented:
> Would this work OK with very large hashes (e.g., 64x64)?

Yes, it seems to keep the advantage on large hashes, too. Converting 3000 randomly generated 64x64 hashes gives the following results:
```
Original version: 15.335927000000083
Updated version: 7.23339599999963
```

It only gives in at about 160-dimensional hashes (25000+ bits). Running 1000 160x160 hashes gives a rough parity:
```
Original version: 37.50718020000204
Updated version: 37.62010579999696
```

bigcat88 commented:
Can I afford another method of conversion?
